### PR TITLE
HHH-16672 Locking entities created in the same transaction fails

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16672

### CODE REVIEW

**Review by beikov — APPROVED** [review]

**Comment by dreab8** [comment]
An alternative solution https://github.com/hibernate/hibernate-orm/pull/7951

File: hibernate-core/src/main/java/org/hibernate/dialect/lock/PessimisticReadSelectLockingStrategy.java [hunk 1]
```diff
@@ -57,6 +57,7 @@ public void lock(Object id, Object version, Object object, int timeout, EventSou
 		final String sql = determineSql( timeout );
 		final SessionFactoryImplementor factory = session.getFactory();
 		try {
+			flushIfRequired( object, session );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/lock/PessimisticReadSelectLockingStrategy.java:60]: IMO the call to `flushIfRequired` should be moved to either `org.hibernate.loader.ast.internal.LoaderHelper#upgradeLock` or `org.hibernate.event.internal.DefaultLockEventListener#onLock` where we have the `EntityEntry` already available.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/dialect/lock/PessimisticReadSelectLockingStrategy.java]: changed it, WDYT @beikov ?

File: hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderHelper.java [hunk 2]
```diff
@@ -90,6 +90,9 @@ public static void upgradeLock(Object object, EntityEntry entry, LockOptions loc
 					entry.forceLocked( object, nextVersion );
 				}
 				else {
+					if ( !entry.isExistsInDatabase() ) {
+						session.forceFlush( entry );
+					}
 					persister.lock( entry.getId(), entry.getVersion(), object, lockOptions, session );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderHelper.java:96]: Maybe I'm missing something, but if the `forceFlush` succeeds i.e. does an insert that doesn't fail, it is guaranteed that this transaction by definition already has an exclusive lock on the inserted row, so AFAICT, there is no need to run the locking select.

```suggestion
					if ( entry.isExistsInDatabase() ) {
						persister.lock( entry.getId(), entry.getVersion(), object, lockOptions, session );
					}
					else {
						session.forceFlush( entry );
					}
```

**Review by beikov — APPROVED** [review]

**Pull request merged**